webrtc_models/__init__.py

```diff
@@ -48,11 +50,158 @@ class RTCConfiguration(_RTCBaseModel):
+        init=False,
+    )
+    related_port: str | None = field(
+        metadata=field_options(alias="relatedPort", serialize="omit"),
+        default=None,
+        init=False,
+    )
+
+    relay_protocol: str | None = field(
+        metadata=field_options(alias="relayProtocol", serialize="omit"),
+        default=None,
+        init=False,
+    )
+    url: str | None = field(
+        default=None, init=False, metadata=field_options(serialize="omit")
+    )
+
+    @classmethod
+    def __pre_deserialize__(cls, d: dict[Any, Any]) -> dict[Any, Any]:
+        """Per spec candidate serializes and deserialises to/from RTCIceCAndidateInit.
+
+        To invoke the toJSON() operation of the RTCIceCandidate interface,
+        run the following steps:
+        1. Let json be a new RTCIceCandidateInit dictionary.
+        2. For each attribute identifier attr in candidate, sdpMid,
+            sdpMLineIndex, usernameFragment:
+        3. Let value be the result of getting the underlying value of the attribute
+            identified by attr, given this RTCIceCandidate object.
+        4. Set json[attr] to value.
+        5. Return json.
+
+        This method reverses that logic.
+        """
+        candidate_init = RTCIceCandidateInit.from_dict(d)
+        return {"candidate_init": candidate_init.to_dict()}
+
+    def __post_init__(self) -> None:
+        """Initialize the class.
+
+        Spec compliance: If both the sdpMid and sdpMLineIndex members of
+        candidateInitDict are null, throw a TypeError.
+        """
+        if (candidate := self.candidate) or (
+            isinstance(self.candidate_init, str) and (candidate := self.candidate_init)
+        ):
+            msg = (
+                "Passing a candidate str to RTCIceCandidate is deprecated, "
+                "pass an RTCIceCandidateInit object"
+            )
+            warn(msg, DeprecationWarning, stacklevel=2)
+            object.__setattr__(
+                self,
+                "candidate_init",
+                RTCIceCandidateInit(candidate, sdp_mid="0"),
+            )
+
+        if TYPE_CHECKING:
+            assert isinstance(self.candidate_init, RTCIceCandidateInit)
+
+        ric_init = self.candidate_init
+        if (
+            ric_init
+            and ric_init.candidate != ""
+            and ric_init.sdp_m_line_index is None
+            and ric_init.sdp_mid is None
+        ):
+            msg = "sdp_m_line_index and sdp_mid cannot both be null."
+            raise TypeError(msg)
+
+        if ric_init:
+            # Attributes are readonly so set them via __setattr__
+            object.__setattr__(self, "candidate", ric_init.candidate)
+            object.__setattr__(self, "sdp_mid", ric_init.sdp_mid)
+            object.__setattr__(self, "sdp_m_line_index", ric_init.sdp_m_line_index)
+            object.__setattr__(self, "user_fragment", ric_init.user_fragment)
+        else:
+            object.__setattr__(self, "candidate", "")
```

Comment:
```suggestion
    candidate: str

    def __post_init__(self) -> None:
        """Initialize class."""
        msg = ("Using RTCIceCandidate is deprecated. Use "
               "RTCIceCandidateInit instead")
        warn(msg, DeprecationWarning, stacklevel=2)

@dataclass(frozen=True)
class RTCIceCandidateInit(RTCIceCandidate):
    """RTC Ice Candidate Init.

    See https://www.w3.org/TR/webrtc/#dom-rtcicecandidateinit
    """

    candidate: str
    sdp_mid: str | None = field(
        metadata=field_options(alias="sdpMid"), default=None, kw_only=True
    )
    sdp_m_line_index: int | None = field(
        metadata=field_options(alias="sdpMLineIndex"), default=None, kw_only=True
    )
    user_fragment: str | None = field(
        metadata=field_options(alias="userFragment"), default=None, kw_only=True
    )

    def __post_init__(self) -> None:
        """Initialize class.

        Spec compliance: If both the sdpMid and sdpMLineIndex members of
        candidateInitDict are null, throw a TypeError.
        """

        if self.sdp_mid is None and self.sdp_m_line_index is None:
            raise TypeError("sdpMid or sdpMLineIndex must be set")

```

Please keep it simple as we deprecate and remove the class anyway.

As `RTCIceCandidateInit` is a subclass `RTCIceCandidate` we can avoid the breaking change. See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides

Changes in core are also "minimal"

```diff
diff --git a/homeassistant/components/camera/__init__.py b/homeassistant/components/camera/__init__.py
index d31d21d424c..121b954ecee 100644
--- a/homeassistant/components/camera/__init__.py
+++ b/homeassistant/components/camera/__init__.py
@@ -20,7 +20,7 @@ from aiohttp import hdrs, web
 import attr
 from propcache import cached_property, under_cached_property
 import voluptuous as vol
-from webrtc_models import RTCIceCandidate, RTCIceServer
+from webrtc_models import RTCIceCandidateInit, RTCIceServer
 
 from homeassistant.components import websocket_api
 from homeassistant.components.http import KEY_AUTHENTICATED, HomeAssistantView
@@ -865,7 +865,7 @@ class Camera(Entity, cached_properties=CACHED_PROPERTIES_WITH_ATTR_):
         return config
 
     async def async_on_webrtc_candidate(
-        self, session_id: str, candidate: RTCIceCandidate
+        self, session_id: str, candidate: RTCIceCandidateInit
     ) -> None:
         """Handle a WebRTC candidate."""
         if self._webrtc_provider:
diff --git a/homeassistant/components/camera/webrtc.py b/homeassistant/components/camera/webrtc.py
index 0612c96e40c..a74b2dfe5ac 100644
--- a/homeassistant/components/camera/webrtc.py
+++ b/homeassistant/components/camera/webrtc.py
@@ -11,7 +11,12 @@ import logging
 from typing import TYPE_CHECKING, Any, Protocol
 
 import voluptuous as vol
-from webrtc_models import RTCConfiguration, RTCIceCandidate, RTCIceServer
+from webrtc_models import (
+    RTCConfiguration,
+    RTCIceCandidate,
+    RTCIceCandidateInit,
+    RTCIceServer,
+)
 
 from homeassistant.components import websocket_api
 from homeassistant.core import HomeAssistant, callback
@@ -78,7 +83,7 @@ class WebRTCAnswer(WebRTCMessage):
 class WebRTCCandidate(WebRTCMessage):
     """WebRTC candidate."""
 
-    candidate: RTCIceCandidate
+    candidate: RTCIceCandidate | RTCIceCandidateInit
 
     def as_dict(self) -> dict[str, Any]:
         """Return a dict representation of the message."""
@@ -146,7 +151,7 @@ class CameraWebRTCProvider(ABC):
 
     @abstractmethod
     async def async_on_webrtc_candidate(
-        self, session_id: str, candidate: RTCIceCandidate
+        self, session_id: str, candidate: RTCIceCandidateInit
     ) -> None:
         """Handle the WebRTC candidate."""
 
@@ -329,7 +334,7 @@ async def ws_candidate(
         return
 
     await camera.async_on_webrtc_candidate(
-        msg["session_id"], RTCIceCandidate(msg["candidate"])
+        msg["session_id"], RTCIceCandidateInit(msg["candidate"], sdp_mid="0")
     )
     connection.send_message(websocket_api.result_message(msg["id"]))
 
diff --git a/tests/components/camera/test_webrtc.py b/tests/components/camera/test_webrtc.py
index ba5cf35c52f..aa3e6f3ada6 100644
--- a/tests/components/camera/test_webrtc.py
+++ b/tests/components/camera/test_webrtc.py
@@ -6,7 +6,7 @@ from typing import Any
 from unittest.mock import AsyncMock, Mock, patch
 
 import pytest
-from webrtc_models import RTCIceCandidate, RTCIceServer
+from webrtc_models import RTCIceCandidate, RTCIceCandidateInit, RTCIceServer
 
 from homeassistant.components.camera import (
     DATA_ICE_SERVERS,
@@ -972,7 +972,7 @@ async def test_ws_webrtc_candidate(
         assert response["type"] == TYPE_RESULT
         assert response["success"]
         mock_on_webrtc_candidate.assert_called_once_with(
-            session_id, RTCIceCandidate(candidate)
+            session_id, RTCIceCandidateInit(candidate, sdp_mid="0")
         )
 
 
@@ -1024,7 +1024,7 @@ async def test_ws_webrtc_candidate_webrtc_provider(
         assert response["type"] == TYPE_RESULT
         assert response["success"]
         mock_on_webrtc_candidate.assert_called_once_with(
-            session_id, RTCIceCandidate(candidate)
+            session_id, RTCIceCandidateInit(candidate, sdp_mid="0")
         )
 
 

 ```